Upgrade to Celery 5.0

king-alexander opened the pull request:
# <!-- Use the title to describe PR changes in the imperative mood --> #

## 🗣 Description ##

<!-- Describe the "what" of your changes in detail. -->
Starting with version 5.0, Celery parses command line options with [Click](https://docs.celeryq.dev/en/stable/history/whatsnew-5.0.html#new-command-line-interface), which introduced breaking changes. I updated the call in `celery.py` that starts the app accordingly.
<!-- To avoid scope creep, limit changes to a single goal. -->
I didn't completely avoid scope creep. In the process of upgrading, I noticed a a minor bug when the component services of admiral were starting up. Redis Commander depends upon Redis, but attempted to start first, leading to a recoverable error. This blip was corrected in `docker-compose.yml`. 

## 💭 Motivation and context ##

<!-- Why is this change required? -->
<!-- What problem does this change solve? How did you solve it? -->
<!-- Mention any related issue(s) here using appropriate keywords such -->
<!-- as "closes" or "resolves" to auto-close them on merge. -->
This change resolves cisagov/admiral#6.   
## 🧪 Testing ##

<!-- How did you test your changes? How could someone else test this PR? -->
<!-- Include details of your testing environment, and the tests you ran to -->
<!-- see how your change affects other areas of the code, etc. -->
I successfully started admiral in both of its modes: `celery-shell` and `bash`. Furthermore, all unit and system tests passed.
<!--
## 📷 Screenshots (if appropriate) ##

Uncomment this section if a screenshot is needed.

-->

## ✅ Pre-approval checklist ##

<!-- Remove any of the following that do not apply. -->
<!-- Draft PRs should have one or more unchecked boxes. -->
<!-- If you're unsure about any of these, don't hesitate to ask. -->
<!-- We're here to help! -->

- [x] This PR has an informative and human-readable title.
- [x] All relevant type-of-change labels have been added.
- [x] I have read the [CONTRIBUTING](../blob/develop/CONTRIBUTING.md) document.
- [x] These code changes follow [cisagov code standards](https://github.com/cisagov/development-guide).
- [x] All relevant repo and/or project documentation has been updated
      to reflect the changes in this PR.
- [x] All new and existing tests pass.

## ✅ Post-merge checklist ##

<!-- Remove any of the following that do not apply. -->

- [ ] Add a tag or create a release.

king-alexander committed:
Upgrade from Celery 4.x

Celery 5.0 introduced breaking changes to the command line interface.
One such change is that you can no longer pass "celery" in argv when
programatically starting the celery app. This commit updates our
invocation and requires Celery 5.0 (Singularity), the lowest stable
version we can support with this change.

dav3r commented on docker-compose.yml:
```diff
@@ -44,6 +44,8 @@ services:
     # build: .
     restart: always
+    depends_on:
+      - redis
```
Since you are touching this file, can you please also alphabetize this YAML (the sections, as well as the options within each section)?  Reviewing is always much easier when things are alphabetical, plus it helps keep us sane.

king-alexander replied:
The build workflow threw errors at me when I changed the order of the top-level sections, but I was otherwise able to alphabetize.

dav3r replied:
The alphabetization looks good to me now, thanks!

dav3r commented on setup.py:
```diff
@@ -91,7 +91,7 @@ def get_version(version_file):
         # Celery 5.0 introduces breaking changes to the CLI. See
         # https://docs.celeryq.dev/en/stable/history/whatsnew-5.0.html?highlight=command%20line%20options#new-command-line-interface
         # for more information.
```
I don't think this comment is still necessary.

dav3r replied:
Thanks for removing that comment in https://github.com/cisagov/admiral/pull/17/commits/92249e56f8c183c64a5ae093b5126994bb8030f4.

dav3r reviewed (changes requested):
Nice work @king-alexander!  I noted some minor changes that I'd like to see included here before I approve.

king-alexander committed:
Alphabetize YAML

dav3r commented on docker-compose.yml:
```diff
@@ -2,88 +2,89 @@
     environment:
-      ADMIRAL_CONFIG_SECTION: cert-worker
-      ADMIRAL_WORKER_NAME: cert
+      - REDIS_HOSTS=default:redis:6379:0:fruitcake
```
Is this environment variable formatted using the `-`/`=` syntax because of the colons in the value?  If so, does this work?  If not, you should add an explanatory comment.
```suggestion
      REDIS_HOSTS: "default:redis:6379:0:fruitcake"
```

dav3r replied:
PS: I realize that you didn't write that line, but I thank you if you're able to clean it up. 😄

king-alexander replied:
That works!

dav3r replied:
Thanks for fixing in https://github.com/cisagov/admiral/pull/17/commits/056fb4c54b0839dcb5e5801ed6bf36dd02f97a88.  Reminder also to please write better commit messages than the default ones that GitHub provides (e.g. "Update docker-compose.yml") when accepting commit suggestions.

dav3r replied:
@king-alexander It looks like the change in https://github.com/cisagov/admiral/commit/056fb4c54b0839dcb5e5801ed6bf36dd02f97a88 has been lost - can you reapply it?

king-alexander replied:
Yes I can. Sorry about that. I must have overwritten some of these changes locally.

mcdonnnj commented on docker-compose.yml:
```diff
@@ -2,88 +2,89 @@
-      - "/run/secrets/redis_conf"
+  celery-flower:
+    environment:
+      # TODO: get this into the secrets raft
```
Another instance of you didn't do this *but*; please make an issue for this TODO and update the comment accordingly.

king-alexander replied:
I updated the TODO comment with [8d0388f](https://github.com/cisagov/admiral/pull/17/commits/8d0388f1ec1dab989beece8a2e9a4220cd83ad6d).

mcdonnnj replied:
Would you please include a brief description for this TODO? Even reusing the original comment that was here is fine. We should just provide context around issues linked so there's some idea of what a TODO is about.

king-alexander replied:
I agree! Addressed with [a9d2916](https://github.com/cisagov/admiral/pull/17/commits/a9d291665037876b243a8173a580ea7f3ee72fb5).

mcdonnnj commented on docker-compose.yml:
```diff
@@ -2,88 +2,89 @@
       ME_CONFIG_MONGODB_ADMINPASSWORD: example
+    image: mongo-express
+    ports:
+      - 8083:8081
```
Per the [documentation](https://docs.docker.com/compose/compose-file/#ports):

> `HOST:CONTAINER` SHOULD always be specified as a (quoted) string, to avoid conflicts with [yaml base-60 float](https://yaml.org/type/float.html).

I know you didn't write it but I just wanted to make sure the why was clearly stated.

```suggestion
      - "8083:8081"
```

dav3r replied:
@king-alexander - I unresolved this conversation because I don't see the ports quoted here.  Not sure if that commit was accidentally removed or what.

king-alexander replied:
Whoops! Thanks for the catch.

mcdonnnj commented on setup.py:
```diff
@@ -88,10 +88,7 @@ def get_version(version_file):
-        # https://docs.celeryq.dev/en/stable/history/whatsnew-5.0.html?highlight=command%20line%20options#new-command-line-interface
-        # for more information.
-        "celery >= 4.3.0,<5.0.0",
+        "celery >= 5.1.0",
```
Just to cover our bases. As an additional question do we *need* `5.1.0` for functionality? We should pin as low as we are able to support whenever possible.

```suggestion
        "celery >= 5.1.0,<6",
```

king-alexander replied:
I read you. To answer your question: no we don't. The lowest possible version we can support with the change is `5.0.0` since that version introduced the breaking change. I updated my commit to reflect proper version pinning. Thanks!

mcdonnnj commented on docker-compose.yml:
```diff
@@ -2,88 +2,89 @@
+    environment:
+      # TODO: get this into the secrets raft
+      CELERY_BROKER_URL: "redis://:fruitcake@redis:6379/0"
+    image: crgwbr/docker-celery-flower
```
Would you make an issue to replace this image? The last image pushed was created four years ago (the GitHub repository is equally dead) and is using an old version of [flower](https://pypi.org/project/flower) from 2016.

king-alexander replied:
Done!

dav3r replied:
A few things here:
1) You should include a TODO comment here that includes the URL of the issue you created.
2) Issue #18 should at least mention the current image name (`crgwbr/docker-celery-flower`) and reference this conversation for context.
3) Whoever starts the conversation in a PR (@mcdonnnj in this case) should be the one to resolve it (except in cases where you accept a commit suggestion and GitHub automatically resolves the conversation).  This gives the original author a chance to verify if the changes are satisfactory to them.
4) You should always include the URL of the commit where you made changes in your response to the conversation.  In this case, that would mean mentioning that you created #18.

king-alexander replied:
Thanks Dave! I re-added a TODO comment with the URL ([8d0388f](https://github.com/cisagov/admiral/pull/17/commits/8d0388f1ec1dab989beece8a2e9a4220cd83ad6d)) and updated the description for #18.

dav3r replied:
Perfect, thanks @king-alexander - that all looks great!

mcdonnnj replied:
Please add an appropriate comment with short description and link to #18 here. You added a link for #19 in the TODO about secrets (above) but have not added anything for updating the `crgwbr/docker-celery-flower` Docker image (#18).

king-alexander replied:
Comment added with [a9d2916](https://github.com/cisagov/admiral/pull/17/commits/a9d291665037876b243a8173a580ea7f3ee72fb5).

mcdonnnj commented on docker-compose.yml:
```diff
@@ -2,88 +2,89 @@
       MONGO_INITDB_DATABASE: certs
       MONGO_INITDB_ROOT_USERNAME: root
       MONGO_INITDB_ROOT_PASSWORD_FILE: /run/secrets/mongo_root_passwd_txt
+    image: mongo
```
Would you please tag this to at least the `major.minor` version that you have verified functionality with in testing.

king-alexander replied:
Updated image tagging with [6c42c08](https://github.com/cisagov/admiral/pull/17/commits/6c42c08f6879b1004e225ff73926a5a38a01cf10).

mcdonnnj commented on docker-compose.yml:
```diff
@@ -2,88 +2,89 @@
       ME_CONFIG_MONGODB_SERVER: mongo
       ME_CONFIG_MONGODB_ADMINUSERNAME: root
       ME_CONFIG_MONGODB_ADMINPASSWORD: example
+    image: mongo-express
```
Same comment about image tag pinning.

king-alexander replied:
Addressed with [6c42c08](https://github.com/cisagov/admiral/pull/17/commits/6c42c08f6879b1004e225ff73926a5a38a01cf10).

mcdonnnj commented on docker-compose.yml:
```diff
@@ -2,88 +2,89 @@
+    command:
+      - "redis-server"
+      - "/run/secrets/redis_conf"
+    image: "redis:alpine"
```
Would you please update this to the `major.minor-alpine` tag aligned with your testing?

king-alexander replied:
Also addressed with [6c42c08](https://github.com/cisagov/admiral/pull/17/commits/6c42c08f6879b1004e225ff73926a5a38a01cf10).

mcdonnnj commented on docker-compose.yml:
```diff
@@ -2,88 +2,89 @@
-      ADMIRAL_WORKER_NAME: cert
+      - REDIS_HOSTS=default:redis:6379:0:fruitcake
+    hostname: redis-commander
+    image: rediscommander/redis-commander:latest
```
Would you please verify that this change works? The `rediscommander/redis-commander` image hasn't been updated in over a year and after checking the [project repository](https://github.com/joeferner/redis-commander) it appears that they are only generating GitHub Container Registry images for now per https://github.com/joeferner/redis-commander/issues/365#issuecomment-1059303415.

```suggestion
    image: ghcr.io/joeferner/redis-commander:0.8
```

king-alexander replied:
Yep, I can confirm it works. I changed the image in [6c42c08](https://github.com/cisagov/admiral/pull/17/commits/6c42c08f6879b1004e225ff73926a5a38a01cf10).

mcdonnnj reviewed (changes requested):
Thanks for figuring out how to get us onto the latest major version! I have some housekeeping and version pinning related feedback for you below.

king-alexander committed:
Update docker-compose.yml

Co-authored-by: Nick <50747025+mcdonnnj@users.noreply.github.com>

king-alexander committed:
Update image tagging

king-alexander committed:
Merge branch 'improvement/upgrade-celery' of https://github.com/king-alexander/admiral into improvement/upgrade-celery

king-alexander committed:
Remove comment

king-alexander committed:
Add TODO for Issue#19

dav3r commented on docker-compose.yml:
```diff
@@ -2,88 +2,89 @@
+    command:
+      - "redis-server"
+      - "/run/secrets/redis_conf"
+    image: "redis:7.0-alpine"
```
We don't need the quotes here:
```suggestion
    image: redis:7.0-alpine
```

king-alexander committed:
Add quotations to Mongo Express ports

The ports in use should be specified as a quoted string to avoid conflicts with YAML Base-60 floating point numbers.

Co-authored-by: Nick <50747025+mcdonnnj@users.noreply.github.com>

king-alexander committed:
Remove quotation marks from Redis Image specification

Co-authored-by: dav3r <daver@geekpad.com>

king-alexander committed:
Change environment variable syntax

Quotation marks around environment variables are preferred to the -/= syntax for consistency with the rest of the file.

Co-authored-by: dav3r <daver@geekpad.com>

dav3r reviewed (approved):
I think everything looks good to me now, thank you @king-alexander for sorting all of this out! ⚓

jsf9k reviewed (approved):
LGTM!

king-alexander committed:
Add TODO for Issue#18

king-alexander committed:
Merge branch 'improvement/upgrade-celery' of https://github.com/king-alexander/admiral into improvement/upgrade-celery

mcdonnnj commented on docker-compose.yml:
```diff
@@ -2,88 +2,92 @@
+      # TODO: https://github.com/cisagov/admiral/issues/19
+      # move into the secrets raft
+      CELERY_BROKER_URL: "redis://:fruitcake@redis:6379/0"
+      # TODO: https://github.com/cisagov/admiral/issues/18
+      # replace this outdated image
+    image: crgwbr/docker-celery-flower
```
Just fixing the alignment.

```suggestion
    # TODO: https://github.com/cisagov/admiral/issues/18
    # replace this outdated image
    image: crgwbr/docker-celery-flower
```

mcdonnnj reviewed (approved):
Thanks for your work on getting this upgrade across the finish line! One small suggestion just to fix a comment's alignment but otherwise LGTM ✔

king-alexander committed:
Realign comment addressing issue #18

Co-authored-by: Nick <50747025+mcdonnnj@users.noreply.github.com>

king-alexander committed:
Merge branch 'develop' into improvement/upgrade-celery

king-alexander committed:
Bump version from 0.2.0 to 1.0.0
